### PR TITLE
Prepare test_autograd.py for introduction of scalars

### PR DESCRIPTION
**Pull request by gchanan**
Distinguish between a few usages of scalar in test_autograd:
1) scalar -> pyscalar when referring to a python number
2) scalar -> 1_element when referring to size {1} tensor
3) () -> NO_ARGS when specifying no arguments to a function instead of a size, since () is now the size of a scalar

Also removes some dead code from when mul_(...) only worked on tensors.

### CODE REVIEW

File: test/test_autograd.py [hunk 1]
```diff
@@ -2512,7 +2513,10 @@ def run_functional_checks(test_case, test_name, name, apply_fn, run_grad_checks,
 
     for dim_perm in product([-1, 1], repeat=len(dim_args_idx)):
         test_name = basic_test_name
-        new_args = [arg * dim_perm[dim_args_idx.index(i)] if i in dim_args_idx else arg for i, arg in enumerate(args)]
+        if args == NO_ARGS:
+            new_args = []
+        else:
+            new_args = [arg * dim_perm[dim_args_idx.index(i)] if i in dim_args_idx else arg for i, arg in enumerate(args)]
```
Comment (apaszke) [inline test/test_autograd.py]: Is this case necessary? It should be an empty list comprehension if `args` is an empty tuple

**Review by apaszke — APPROVED** [review]

File: test/test_autograd.py (hunk 1 above)
Comment (gchanan) [reply test/test_autograd.py]: when args == NO_ARGS enumerate doesn't work on it; I can add an iterator that immediately exits.

Comment (apaszke) [reply test/test_autograd.py]: What's the problem? I just tried enumerating an empty sequence and it was ok

Comment (gchanan) [reply test/test_autograd.py]: it's not an empty sequence, it's an object().

**Comment by gchanan** [comment]
trusty build failure doesn't look related.

**Pull request merged**